Woop: store onboarding steps data

retrofox opened the pull request:
~### It depends on https://github.com/Automattic/wp-calypso/pull/59785~

#### Changes proposed in this Pull Request

This PR collects and stores the store address into the site options. For this purpose, it implements a new `useSiteOption()` custom hook:

### Only in Development

Check that this `Store Address` step is only visible in development environment.

### Cuctom hook

Ir adds a new custom hook to make the data handling easier.

```es6
import { useSiteOption } from '../hooks/use-site-options';

function StoreAddressFrom() {
  const { get, save, update } = useSiteOptions( 'site-id' );

  return (
    <>
      <TextControl
        label={ __( 'Address line 1' ) }
        value={ get( 'address_1' ) }
        onChange={ ( value ) => update( { address_1: value } ) }
      />

      <Button onClick={ save }>Save</Button>
  </>
  );
}
```

https://user-images.githubusercontent.com/77539/148269203-ded15a22-f38c-4ec5-9d96-99c5b571fc28.mov

#### Testing instructions

* Apply testing patch D72185-code
<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Test on development environment
* Fill address data
* Click on `Confirm` button. You can confirm the client performs a POST request in order to save the data
* Confirm the data is persistent, basically via a hard-refresh.
* Confirm next step is `Confirm` step
* Check the back button in the `Confirm` step. It should lead to the `Store Address` step.

Check the feature is not visible on other envs.
You can use the Calypso live branch:
`https://container-flamboyant-banach.calypso.live/woocommerce-installation/<your-testing-site>`

* It should not show the Store Address step.
* Check the back button in the Confirm step. It should lead to the landing page.

<img width="975" alt="Screen Shot 2022-01-06 at 5 26 26 PM" src="https://user-images.githubusercontent.com/1689238/148461100-b28df0ee-eb39-44c9-865b-23997075bc2c.png">

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to
* https://github.com/Automattic/wp-calypso/issues/57423
* https://github.com/Automattic/wp-calypso/issues/59481

retrofox committed:
woop: ad useSiteOptions() custom hook

retrofox committed:
woop: add pretty simple store-address step cmp

retrofox committed:
woop: register store-address step

retrofox committed:
woop: change back URL for confirm step

lsl commented:
This looks pretty promising :+1:

retrofox committed:
woop: update site settings properly

retrofox committed:
woop: fix store-address ui

retrofox committed:
woop: tidy use-site-options hook

retrofox committed:
woop: improve updating site options data

retrofox committed:
woop: remove site option when empty

retrofox committed:
woop: do not expose data prop in the hook

retrofox committed:
woop: tweak styles

retrofox committed:
woop: re implement useSiteOption for single options

retrofox committed:
woop: first store-address super basic implementation

retrofox committed:
woop: update doc

retrofox committed:
signup: fix store address folder path

obenland commented on client/signup/steps/woocommerce-install/hooks/use-site-options/index.tsx:
```diff
@@ -0,0 +1,105 @@
+import { getSiteSettings } from 'calypso/state/site-settings/selectors';
+
+// WooCommerce single options.
+export const WOOP_STORE_ADDRESS_1 = 'woocommerce_store_address';
```
Anything standing in the way of using a `WOOCOMMERCE_` prefix over `WOOP_`? With Woop being a temporary project it might not be super easy to understand what it references two or three years from now

retrofox replied:
https://github.com/Automattic/wp-calypso/pull/59505/commits/4a46d2efef52e399b1119ff84da332d4656e1c12

obenland commented on client/signup/steps/woocommerce-install/hooks/use-site-options/index.tsx:
```diff
@@ -0,0 +1,105 @@
+ * @param {number} siteId - site id
+ * @returns {Array} - site option handlers
+ */
+export function useSiteOption( siteId: number ) {
```
It looks like Calypso uses the `SiteSettings` terminology. Could we continue using that here? I'm worried about introducing ambiguity by using two terms for the same thing.

retrofox replied:
renaming to `useSiteSettingsOption()` in case you agree.

obenland replied:
Can you help me understand why it's important to keep the `Option` term?

retrofox replied:
SGTM

obenland commented on client/signup/steps/woocommerce-install/hooks/use-site-options/index.tsx:
```diff
@@ -0,0 +1,105 @@
+		dispatch( saveSiteSettings( siteId, editedOptions ) ).then( () => setEditedOptions( {} ) );
+	}, [ dispatch, editedOptions, siteId ] );
+
+	return { save, update, get };
```
Building on the `SiteSetting` thought: Have you considered calling these functions `saveSiteSetting `, `updateSiteSetting()`, and `getSiteSetting()`? 
Looking at how they're used later on, save/update/get feel a bit ambiguous, especially once this hook gets used outside of our component.

retrofox replied:
from my PoV, the ambiguity disappears when it calls the hook:

```es6
const { get, save, update } = useSiteSettingsOption( siteId );
```

I prefer shorter names. It's pretty simple to rename them, too:

```es6
const {
  get: getSiteSetting,
  save: saveSiteSetting,
  update: updateSiteSetting,
} = useSiteSettingsOption( siteId );
```

we could force to define the method names returning an array instead of an object:

```es6
// define shorter names. Order matters.
const [ get, update, save ] = useSiteSettingsOption( siteId );

// more descriptive names. Order matters.
const [ getSiteSetting, updateSiteSetting , saveSiteSetting ] = useSiteSettingsOption( siteId );
```

Totally open to do any change.

obenland replied:
@allilevine @DustyReagan @Rolilink Any thoughts here?

allilevine replied:
Looks like I'm late to this discussion 😅 More descriptive names are clearer, but `saveSiteSetting` is very close to the action name, `saveSiteSettings`. Is there an alternative?

DustyReagan replied:
I don't have a strong opinion. Usually if I'm in doubt about name I lean more verbose. I think Allison has a good point about `saveSiteSetting` vs `saveSiteSettings`.

Rolilink replied:
Also late here but I lean more toward `saveSiteSetting`, also there is a value maybe into naming actions more verbose like `saveSiteSettingsAction`?

retrofox replied:
something to keep in mind, for the next iterations, is the ability to deal with site options that contain only one value, for instance, a String or Number. This is the case for these new options: `woocommerce_store_address`, `woocommerce_store_address_2`, etc...

However, thinking in advance, we'll have to deal with storing data into an option that could be an Object/Array. This is the case for the onboarding profile option. It's an object, where nested fields also can be an Object/Array. The idea of adding this new hook is, also, being an abstraction layer to simplify the data handling into the components.

Then, probably, we'll have to define two hooks or check the type of the value in order to define how to store it into the site setting option. I hope it makes sense for you :intense:

retrofox replied:
My opinion is to keep the API as it, because:

* The fact the hook returning an object instead of an array allows defining what are the methods what we want to consume clearer:
```es6
const { get, save } = useSiteSettingsOption( siteId );
```
vs
```es6
// it is not handy
const [ get, update, ...save ] = useSiteSettingsOption( siteId );
```

* The order doesn't matter
```es6
const { get, save } = useSiteSettingsOption( siteId );
```

```es6
const { save, get } = useSiteSettingsOption( siteId );
```

* It's pretty simple to rename the method with more descriptive names:
```es6
const {
  get: getSiteSetting,
  save: saveSiteSetting,
  update: updateSiteSetting,
} = useSiteSettingsOption( siteId );
```

retrofox committed:
woop: unchain cleaning local store when save

retrofox committed:
woop: rename constant names

retrofox committed:
woop: rename hook to useSiteSettings

obenland commented on client/signup/config/steps-pure.js:
```diff
@@ -808,6 +808,17 @@ export function generateSteps( {
+		'store-address': {
+			stepName: 'store-address',
+			props: {
+				headerTitle: i18n.translate( 'Add an adress so you can get paid' ),
```
@SaxonF I understand that this is the wording as shown in the figma file. I'm curious what you think about "Add an address to accept payments" to make it a little less blunt? Or something along those lines?
cc @ianstewart

SaxonF replied:
I like that better @obenland 👍

DustyReagan commented on client/signup/steps/woocommerce-install/hooks/use-site-settings/Readme.md:
````diff
@@ -0,0 +1,46 @@
+```es6
+import useSiteSettings from './use-site-settings';
+
+```es6
````
```suggestion
```

Rolilink commented on client/signup/steps/woocommerce-install/hooks/use-site-settings/index.tsx:
```diff
@@ -0,0 +1,109 @@
+
+	/*
+	 * Helper to 'update' site settings data.
+	 * Chnages are applied to the Redux store.
```
`changes`

retrofox replied:
addressed here https://github.com/Automattic/wp-calypso/pull/59505/commits/eaa1fc957a1ffac40d849fece0f71a3facb7dd27. Thanks

Rolilink commented on client/signup/steps/woocommerce-install/hooks/use-site-settings/index.tsx:
```diff
@@ -0,0 +1,109 @@
+	const update = useCallback(
+		( option: string, value: string ) => {
+			const siteOption = getSiteOptionName( option );
+			setEditedSettings( ( state ) => ( { ...state, [ siteOption ]: value } ) );
```
Why are this settings being saved in the redux store and also a private store? any particular reason why the redux store can't be the single source of truth?

retrofox replied:
Nice question, Rolando.
We need to track what are the options that we are going to ~update~ save.
If the user fills the address 1, address, 2, city... etc, we need to collect them in order to pass what are the final values that we want to save.
The [update](https://github.com/Automattic/wp-calypso/blob/5166a2a25e5ce8b94e882dd0a32847f687cb86b5/client/state/site-settings/actions.js#L39) method only updates the redux store, while the [save](https://github.com/Automattic/wp-calypso/blob/5166a2a25e5ce8b94e882dd0a32847f687cb86b5/client/state/site-settings/actions.js#L39) method performs the POST endpoint request.
But the [save method](https://github.com/Automattic/wp-calypso/blob/5166a2a25e5ce8b94e882dd0a32847f687cb86b5/client/state/site-settings/actions.js#L86) ALSO expects a settings object as the second argument, and that is the private store (useState) that we pass to it.
In an Ideal World, we shouldn't take care about what are the data that we want to save since it should be tracked by redux when the `update` method is called. We simply should call `save()` without passing any settings data.

Rolilink replied:
Makes sense, I wonder if instead we can use the private store to just track which attributes are being updated (address 1, address 2) as a string and then pick them up from the redux state. Maybe it can be a more descriptive way of declaring the private store?
```
// on declaration
const [ editedSettings, setEditedSettings ] = useState([]);

// on update
setEditedSettings([ ...editedSettings,  'address1']);

// on save
const settingsToSave = editedSettings.map(key => settings[key]);
```

retrofox replied:
that makes totally sense. I'd suggest defining a function to deal with updating the state in the hook, instead of calling it value. Something like:

```es6
setEditedSettings( ( settings ) => ( [ ...settings,  'address1'] ) );
```

Also, check we don't duplicate items into the Array.

Rolilink replied:
Good catch! there! we totally need to check for uniqueness too!

Rolilink replied:
Addressed here: https://github.com/Automattic/wp-calypso/pull/59505/commits/0621f7369c2e38e374f3e3d0cf499e59053601f3

retrofox committed:
woop: refactor code in order to fix TS type issues

retrofox committed:
selectors: add more keys to SiteSettingsItem type

retrofox committed:
Update client/signup/steps/woocommerce-install/hooks/use-site-settings/Readme.md

Co-authored-by: Dusty Reagan <dusty@dustyreagan.com>

retrofox committed:
woop: fix typo

retrofox committed:
woop: minor useSiteSettings doc improvement

retrofox commented on client/state/site-settings/selectors.js:
```diff
@@ -38,6 +38,11 @@ export function getSiteSettingsSaveRequestStatus( state, siteId ) {
 /**
  * @typedef {{
  blog_public?: number;
+ woocommerce_store_address?: string;
+ woocommerce_store_address_2?: string;
+ woocommerce_store_city?: string;
+ woocommerce_store_postcode?: string;
+ woocommerce_default_country?: string;
  }} SiteSettingsItem
```
we need to register these new fields to fix TS type issues.

DustyReagan commented on client/signup/config/steps-pure.js:
```diff
@@ -808,6 +808,17 @@ export function generateSteps( {
+		'store-address': {
+			stepName: 'store-address',
+			props: {
+				headerTitle: i18n.translate( 'Add an adress so you can get paid' ),
```
```suggestion
				headerTitle: i18n.translate( 'Add an address so you can get paid' ),
```
Typo fix, in case it stays the same.

DustyReagan commented on client/signup/config/steps-pure.js:
```diff
@@ -808,6 +808,17 @@ export function generateSteps( {
+			props: {
+				headerTitle: i18n.translate( 'Add an adress so you can get paid' ),
+				headerDescription: i18n.translate(
+					'This will used as your default business address. You can change it later if you need to.'
```
```suggestion
					'This will be used as your default business address. You can change it later if you need to.'
```

retrofox committed:
Update client/signup/config/steps-pure.js

Co-authored-by: Dusty Reagan <dusty@dustyreagan.com>

retrofox committed:
Update client/signup/config/steps-pure.js

Co-authored-by: Dusty Reagan <dusty@dustyreagan.com>

retrofox committed:
woop: set back URL in confirm step depending on feature flag

Rolilink committed:
woop: private store in useSiteSettings track only changes
- Private store is now tracking changes as an array of attributes, this helps us have the redux store as the SST.

retrofox committed:
woop: add store-address step to woocommerce installer when the `woop`
feature flag is enabled

retrofox committed:
woop: request countries in custom hook

retrofox committed:
woop: replace text input by countries selector

obenland commented on client/signup/steps/woocommerce-install/step-store-address/index.tsx:
```diff
@@ -0,0 +1,116 @@
+						value={ get( WOOCOMMERCE_DEFAULT_COUNTRY ) }
+						onChange={ handleCountryChange }
+						countriesList={ countriesList }
+						translate={ __ }
```
Does `FormCountrySelect` not get `translate` from [`localize`](https://github.com/Automattic/wp-calypso/blob/171ef42853484467b2bc4cf575547d9e41f8057c/client/components/forms/form-country-select/index.jsx#L67)?

retrofox replied:
~[nop, it doesn't](https://github.com/Automattic/wp-calypso/blob/40853249648188a38f4ccbdb205f8e2b108a7c7f/client/components/forms/form-country-select/index.jsx#L16). The `translate` is a component property :-/~

retrofox replied:
let me check... 😓

retrofox replied:
It's weird. When I remove the prop I get the error:

![image](https://user-images.githubusercontent.com/77539/148443446-50afadbc-62e7-4132-a388-aa4f426a5c37.png)

retrofox replied:
Addressed [here](https://github.com/Automattic/wp-calypso/pull/59505/commits/71c78d69e64f442cab288f23e7c63296bea28bf0). Sorry, and thanks!

retrofox committed:
woop: use default from country selector cmp

retrofox committed:
woop: fix Readme doc format issue

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7076112

Hi @retrofox, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Add an address so you can get paid`
- `This will be used as your default business address. You can change it later if you need to.`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7076112).
